src/main/java/edu/kit/ipd/crowdcontrol/objectservice/moneytransfer/MailParser.java

```diff
@@ -0,0 +1,101 @@
+ */
+public class MailParser {
+
+    /**
+     * Extracts a giftcode out of a message.
+     *
+     * @param msg the message, to extract the code
+     * @return returns the giftcode
+     * @throws MoneyTransferException gets thrown, if an error occurred during parsing
+     */
+    protected static Optional<GiftCodeRecord> parseAmazonGiftCode(Message msg) throws MoneyTransferException {
+        //Extract Message
+        String message;
+        Optional<GiftCodeRecord> optional;
+        try {
+            if (!msg.getFrom()[0].toString().toLowerCase().endsWith("amazon.de>")) {
+                return Optional.ofNullable(null);
+            }
+        } catch (MessagingException e) {
+            return Optional.ofNullable(null);
+        }
+        try {
+            Multipart parts = (Multipart) msg.getContent();
+            BodyPart body = parts.getBodyPart(0);
+            Multipart innerMsg = (Multipart) body.getContent();
+            BodyPart textBody = innerMsg.getBodyPart(0);
+            message = textBody.getContent().toString();
+        } catch (ClassCastException | MessagingException | IOException e) {
+            throw new MoneyTransferException("The Parser cannot extract the giftcode from the mails, because the mail format changed. You need to adjust the parser to the new mail format.");
+        }
+        //Parse Message
+        String messageStr = message.replaceAll(" ", "");
+
+        StringBuilder password = MoneyTransferManager.loadMessage("src/main/resources/parsingPassword.txt");
+        if (!messageStr.contains(password)) {
+            return Optional.ofNullable(null);
+        }
+
+        String codePatternStr = "[0-9A-Z]+(-[0-9A-Z]+)+";
+        Pattern codePattern = Pattern.compile(codePatternStr);
+        Matcher codeMatcher = codePattern.matcher(messageStr);
+
+        String amountPatternStr = "[0-9]+[,|.][0-9][0-9][€]";
+        Pattern amountPattern = Pattern.compile(amountPatternStr);
+        Matcher amountMatcher = amountPattern.matcher(messageStr);
+
+        if (!checkMatches(codeMatcher) || !checkMatches(amountMatcher)) {
+            throw new MoneyTransferException("The Parser cannot extract the giftcode from the mails, because the mail format changed. You need to adjust the parser to the new mail format.");
+        }
+
+        String giftCode = codeMatcher.group(0);
+
+        String amountStr = amountMatcher.group(0);
+        amountStr = amountStr.replaceAll(",", "").replaceAll("\\.", "").replaceAll("€", "");
+
+        GiftCodeRecord rec = new GiftCodeRecord();
+        rec.setAmount(Integer.parseInt(amountStr));
+        rec.setCode(giftCode);
+
+        return Optional.of(rec);
+    }
+
+    /**
+     * Checks, whether all groups of a matcher are identical.
+     *
+     * @param matcher the matcher to check
+     * @return returns, whether the groups are identical
+     */
+    private static boolean checkMatches(Matcher matcher) {
```

Comment:
additional check, that i dont have more than 1 string in the mail, that match my giftcode/amount pattern, so i dont save wrong giftcodes.